cpp/include/cudf/aggregation.hpp

```diff
@@ -599,6 +609,244 @@ std::unique_ptr<Base> make_udf_aggregation(udf_type type,
                                            std::string const& user_defined_aggregator,
                                            data_type output_type);
 
+/**
+ * @brief The interface for host-based UDF implementation.
+ *
+ * An implementation of host-based UDF needs to be derived from this base class, defining
+ * its own version of the required functions.
+ */
+struct host_udf_base {
+  host_udf_base()          = default;
+  virtual ~host_udf_base() = default;
+
+  /**
+   * @brief Define the possible data needed for groupby aggregations.
+   *
+   * Note that only sort-based groupby aggregations are supported.
+   */
+  enum class groupby_data_attribute : int32_t {
```

Comment:
Nit: 
Add attribute `num_groups`, some code invoke: `helper.num_groups(stream)`
We can get `num_groups` via:
```
offsets.size() - 1
```
Not sure the perf/mem diff between the two approaches

Reply:
That's equivalent to accessing `GROUP_OFFSETS`. Anyway, getting `NUM_GROUPS` directly would be more straightforward. Will add such data attribute.